classes/plugins/PKPPubIdPlugin.inc.php

```diff
@@ -0,0 +1,410 @@
+		if (!parent::register($category, $path)) return false;
+		// Enable storage of additional fields.
+		foreach($this->getDAOs() as $publicObjectType => $daoName) {
+			HookRegistry::register(strtolower_codesafe($daoName).'::getAdditionalFieldNames', array($this, 'getAdditionalFieldNames'));
+		}
+		return true;
+	}
+
+	/**
+	 * @copydoc Plugin::getActions()
+	 */
+	function getActions($request, $actionArgs) {
+		$router = $request->getRouter();
+		import('lib.pkp.classes.linkAction.request.AjaxModal');
+		return array_merge(
+				$this->getEnabled()?array(
+						new LinkAction(
+								'settings',
+								new AjaxModal(
+										$router->url($request, null, null, 'manage', null, $actionArgs),
+										$this->getDisplayName()
+										),
+								__('manager.plugins.settings'),
+								null
+								),
+				):array(),
+				parent::getActions($request, $actionArgs)
+				);
+	}
+
+ 	/**
+	 * @copydoc Plugin::manage()
+	 */
+	function manage($args, $request) {
+		$user = $request->getUser();
+		$router = $request->getRouter();
+		$context = $router->getContext($request);
+
+		$settingsFormName = $this->getSettingsFormName();
+		$settingsFormNameParts = explode('.', $settingsFormName);
+		$settingsFormClassName = array_pop($settingsFormNameParts);
+		$this->import($settingsFormName);
+		$form = new $settingsFormClassName($this, $context->getId());
+		$notificationManager = new NotificationManager();
+		if ($request->getUserVar('save')) {
+			$form->readInputData();
+			if ($form->validate()) {
+				$form->execute();
+				$notificationManager->createTrivialNotification($user->getId(), NOTIFICATION_TYPE_SUCCESS);
+				return new JSONMessage(true);
+			} else {
+				return new JSONMessage(true, $form->fetch($request));
+			}
+		} elseif ($request->getUserVar('clearPubIds')) {
+			$contextDao = Application::getContextDAO();
+			$contextDao->deleteAllPubIds($context->getId(), $this->getPubIdType());
+			$notificationManager->createTrivialNotification($user->getId(), NOTIFICATION_TYPE_SUCCESS);
+			return new JSONMessage(true);
+		} else {
+			$form->initData();
+			return new JSONMessage(true, $form->fetch($request));
+		}
+	}
+
+
+	//
+	// Protected template methods to be implemented by sub-classes.
+	//
+	/**
+	 * Get the public identifier.
+	 * @param $pubObject object
+	 * 	OJS Article, Issue, or ArticleGalley i.e. OMP PublicationFormat
+	 * @param $preview boolean
+	 *  when true, the public identifier will not be stored
+	 * @return string
+	 */
+	abstract function getPubId($pubObject, $preview = false);
+
+	/**
+	 * Construct the public identifier from its prefix and suffix.
+	 * @param $pubIdPrefix string
+	 * @param $pubIdSuffix string
+	 * @param $contextId integer
+	 * @return string
+	 */
+	abstract function constructPubId($pubIdPrefix, $pubIdSuffix, $contextId);
+
+	/**
+	 * Public identifier type, see
+	 * http://dtd.nlm.nih.gov/publishing/tag-library/n-4zh0.html
+	 * @return string
+	 */
+	abstract function getPubIdType();
```

Comment:
Nice to finally be able to use abstract functions instead of

    fatalError('WHY ARE WE HANDLING ABSTRACTS THIS WAY?!');

:)